R0204: Redefinition of <name> type - False positive on async functions.

### Bug description

Issue was already identified here: https://github.com/PyCQA/pylint/issues/1279 - The issue was closed, but the bug is still there.

This will yield a false negative:

```python
class Root:
    ...


class TestTest:
    async def test_one(self):
        potato = Root()
        assert potato

    async def test_two(self):
        potato = {}
        assert potato
```

### Configuration

```ini
[tool.pylint.MASTER]
disable = [
    "C0114",
    "C0115",
    "C0116",
    "E0213",
    "R0801",
    "R0901",
    "R0911",
    "R0913",
    "R6103",
    "W0201",
    "W0238",
    "W0212",
    "W0613",
]
max-line-length = 88
extension-pkg-allow-list="pydantic"
min-public-methods = 0
default-docstring-type = "google"
string-quote="double"
triple-quote="double"
docstring-quote="double"
load-plugins = [
    "pylint.extensions.bad_builtin",
    "pylint.extensions.broad_try_clause",
    "pylint.extensions.check_elif",
    "pylint.extensions.code_style",
    "pylint.extensions.comparison_placement",
    "pylint.extensions.confusing_elif",
    "pylint.extensions.consider_ternary_expression",
    "pylint.extensions.docparams",
    "pylint.extensions.empty_comment",
    "pylint.extensions.eq_without_hash",
    "pylint.extensions.for_any_all",
    "pylint.extensions.overlapping_exceptions",
    "pylint.extensions.private_import",
    "pylint.extensions.redefined_variable_type",
    "pylint.extensions.set_membership",
    "pylint_quotes",
]
```


### Command used

```shell
pylint a.py
```


### Pylint output

```shell
************* Module a
a.py:11:8: R0204: Redefinition of potato type from a.Root to dict (redefined-variable-type)

------------------------------------------------------------------
Your code has been rated at 8.89/10 (previous run: 8.89/10, +0.00)
```


### Expected behavior

Here's the output if the methods are not `async`:

```
class Root:
    ...


class TestTest:
    def test_one(self):
        potato = Root()
        assert potato

    def test_two(self):
        potato = {}
        assert potato
```

```
pylint a.py 

-------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 8.89/10, +1.11)
```

### Pylint version

```shell
2.15.10
```


### OS / Environment

Linux, Debian sid

### Additional dependencies

_No response_

--- a/doc/whatsnew/fragments/8120.false_positive
+++ b/doc/whatsnew/fragments/8120.false_positive
@@ -1,0 +1,3 @@
+Fix a false positive for ``redefined-variable-type`` when ``async`` methods are present.
+
+Closes #8120

--- a/pylint/extensions/redefined_variable_type.py
+++ b/pylint/extensions/redefined_variable_type.py
@@ -46,8 +46,8 @@ class MultipleTypesChecker(BaseChecker):
     def leave_classdef(self, _: nodes.ClassDef) -> None:
         self._check_and_add_messages()
 
-    visit_functiondef = visit_classdef
-    leave_functiondef = leave_module = leave_classdef
+    visit_functiondef = visit_asyncfunctiondef = visit_classdef
+    leave_functiondef = leave_asyncfunctiondef = leave_module = leave_classdef
 
     def visit_module(self, _: nodes.Module) -> None:
         self._assigns: list[dict[str, list[tuple[nodes.Assign, str]]]] = [{}]

--- a/tests/functional/ext/redefined_variable_type/redefined_variable_type.py
+++ b/tests/functional/ext/redefined_variable_type/redefined_variable_type.py
@@ -83,3 +83,28 @@ def func2(x):
     else:
         var4 = 2.
         var4 = 'baz'  # [redefined-variable-type]
+
+
+# Test that ``redefined-variable-type`` is not emitted
+# https://github.com/PyCQA/pylint/issues/8120
+
+async def test_a():
+    data = [
+        {'test': 1},
+        {'test': 2},
+    ]
+    return data
+
+async def test_b():
+    data = {'test': 1}
+    return data
+
+
+class AsyncFunctions:
+    async def funtion1(self):
+        potato = 1
+        print(potato)
+
+    async def funtion2(self):
+        potato = {}
+        print(potato)

